### ISSUE TITLE
Compose Observables instead of updating Atoms in onValue

### ISSUE THREAD

**Issue by raimohanska** [eemelisoderholm/armadillochat#1]
https://github.com/eemelisoderholm/armadillochat/blob/970c74bca28b48e1081b6f4509d0794e52c1b9ff/src/client/state/channels.js#L10

You might try rewriting `channels` using observable composition, such as [`scan`](https://rpominov.github.io/kefir/#scan). In Bacon.js there's also a very handy [`Bacon.update`](https://baconjs.github.io/api.html#bacon-update) method for creating `Properties` based on multiple `EventStreams`.

Just a suggestion :)

**Comment by eemelisoderholm** [comment]
Thanks for the tip!

I tried using `scan` to handle the incoming socket messages kinda like actions in redux, emitting new set of channels in an observable.

Any thoughts on this approach?
https://github.com/eemelisoderholm/armadillochat/blob/d695a3825191d6c3ec686f2c19e796cc8ef840f5/src/client/streams/channels.js

Is the idea of just storing selected ids in a atom and deriving the actual data of selected items on demand through the observables sane? I haven't found many examples about state management with networking yet.

**Comment by raimohanska** [comment]
I think it's a good idea to only store "root data" in Atoms and express all derived data using observable composition, just like you did there with `scan`. This way there's no redundancy and everything keeps in sync nicely.

From performance point of view, I'd re-consider including all the messages in the channels observable though :)

**Comment by eemelisoderholm** [comment]
Agreed, this was more of an ad-hoc way to store the messages initially before private messaging.

Next step in terms of the improving the model would be separating all messages from channels and users into `conversations` stream, with users and channels streams only carrying a reference to their related conversation. This way messages updates can be handled in one place and all the sources can be updated on a single level instead of inserting things further down in their property trees.

In the backend I'd probably set up a database with tables for users, channels, conversations and messages, with one-to-many join table for the latter two. But for now the server is there just to enable practicing things with the client and not a big focus.
